### PR TITLE
🚀 REST API: AGI can Create, Start and Stop multiple Autonomous Agents (Monitoring coming soon)🚀

### PR DESCRIPTION
**Pull request by merwanehamadi**
### Background
The goal is to build an AGI. This AGI might be one single Auto-GPT that loops against itself, but I doubt it because the LLM is super slow. We can't expect to have something that has high computing power with only one autonomous agent. And we can't have one agent per computing machine, that's stupid, because most of the time the running process will just hang and do nothing, waiting for Open AI to give back an answer. Waste of resources. Running multiple agents in different tabs is not ideal either because processes are heavier than threads.
On the current PR I was able to spawn 20 concurrent workers simultaneously without any problem.

So I think that in order to build an AGI, we need to build an entity able to:
- Define autonomous agents.
- Start autonomous agents
- Monitor autonomous agents. This means being able to identify all the meaningful events happening to them and talk to them. (this is not part of this PR unfortunately, it's too many changes at once, but I have a branch ready for it) => 
- Stop these workers when necessary.

Which programmatic language is the AGI going to use to do that? A proper, well-defined REST API is the way to go, we can only do so much with a CLI.

This AGI will talk to its workers through rest api calls, which are nothing more than JSON sent over the network.

At the beginning we will be playing the AGI, so we will use these endpoints to communicate with our agents, through a web browser UI.

### Changes
Now let's come back to reality, I had to work with what the repo provides, so I made the smallest changes possible to achieve these goals. **The first thing I want to say, is that my changes won't affect the users in any way. All the existing features are preserved and unchanged.**

Additions to the repo **that are all necessary to achieve this User Story** (I insist they're all necessary to be merged together):
- A well-structured web server with FastAPI.
- Solid integration tests, extremely readable because all the expected requests and responses are written. These integration tests record the behavior of the autonomous agents. I will use these tests to maintain an automatically generated documentation of auto-gpt api. 
- The integration tests work with GPT3.5 and the cost is very small. Talking about cost, @Torantulino please create an online bank account for auto-gpt organization so we can wire money and pay for an api key that will use to call open AI within github action. You can keep the key private. I will wire 100$ to the bank account when you have it ready. Also, I know some companies interested sponsoring auto-gpt, so reach out to me on discord.
- Ability to stop an autonomous agent gracefully, without forcing it. Each agent runs in separates threads.
- a NoMemory class. Why is that ? Because right now memories are shared accross the same worker !!! So it means if we spawn multiple autonomous agents, they will step into each other's toes. The great news is that NoMemory is actually super performant, I have tested it and it performs oftentimes better than pinecone, mostly because we load an ugly JSON of all the past memories and it confuses the AI. Particularly with GPT 3.5. It turns out summarizing the content as we go and loosing any non essential information can be great for short tasks.
- the server is currently in a folder called app with the right project structure for a project of this size. This was the easiest way to do it without breaking the whole repo. This repo has no module, and I can't put the api under a folder called "scripts", that's beyond me :)

CHANGES to the repo:
- finally wrap the main.py into an "if ___main___", otherwise nothing in this project is reusable, right now I can't import the main.py without starting the loop haha
- create the autonomous agent class so that it can start the start_loop method
- create attributes of the class. It's pretty much adding self everywhere.
- The temperature is currently None in the repo which means 1. The values are between 0 and 2. For my tests I need a very consistent LLM, so I need 0. So I just put an environment variable. The value will still be 1 for others and nothing changes. (I recommend putting the default value to 0 by the way this is a game changer, but that's another story...)

### Documentation
Look at the tests, they document clearly the following requests and responses as well as the output of the worker:
2 integration tests at the moment: 
- Autonomous agent defined (POST), started (PATCH) STOPPED immediately after starting (PATCH)
- Autonomous agent defined (POST), started (PATCH), Writes a file and shuts down on its own. When we try to shut it down we get a 500 because it's already down.
Screenshot 2023-04-10 at 8 32 34 PM
Screenshot 2023-04-10 at 8 32 29 PM
Screenshot 2023-04-10 at 8 32 24 PM
Screenshot 2023-04-10 at 8 32 16 PM

### Test Plan
The CLI has to keep working !!! So I tested it thoroughly Here is one run example.
Screenshot 2023-04-10 at 8 27 02 PM
Screenshot 2023-04-10 at 8 27 12 PM
### PR Quality Checklist
- [X] My pull request is atomic and focuses on a single change.
- [X] I have thouroughly tested my changes with multiple different prompts.
- [X] I have considered potential risks and mitigations for my changes.
- [X] I have documented my changes clearly and comprehensively.
- [X] I have not snuck in any "extra" small tweaks changes

### CODE REVIEW

**Comment by merwanehamadi** [comment]
Hang tight on this one, I am making sure the decorator to stop the threads is added to al the major IO methods.

I know this is going to be a long road to merge this one. Once it's ready to be merged I will create more commits, currently, it's all into one.

File: .gitignore [hunk 1]
```diff
@@ -12,3 +12,12 @@ outputs/*
 ai_settings.yaml
 .vscode
 auto-gpt.json
+.idea
+venv
+.idea
+ai_settings.yaml
+app/.DS_Store
+.DS_Store
```
Comment (dhensen) [inline .gitignore:18]: Good that you added .DS_Store to gitignore, I still see lots of .DS_Store files already committed, please remove them.

File: app/web/api/autonomous_agents/views.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        if "thread_id" not in body:
+            return JSONResponse(content={"message": "thread_id is required in the request body"}, status_code=400)
+        thread_id = body["thread_id"]
+        # Terminate the thread using the thread ID
+        thread = thread_dict.get(thread_id)  # Get the thread from the dictionary
+        if thread:
+            if not thread.is_alive():
+                return JSONResponse(content={"message": "The thread is already terminated"}, status_code=500)
+
+            threads_to_terminate = read_file("threads_to_terminate.json")
```
Comment (dhensen) [inline app/web/api/autonomous_agents/views.py]: reading and json.loading the threads_to_terminate.json could be in a function, it's also used in the `def termination_check`

File: requirements.txt [hunk 3]
```diff
@@ -15,3 +15,8 @@ pinecone-client==2.2.1
 redis
 orjson
 Pillow
+fastapi>=0.94.1,<0.95
+uvicorn[standard]==0.21.0
+pydantic[dotenv]==1.10.6
+yarl==1.8.2
+ujson==5.7.0
```
Comment (dhensen) [inline requirements.txt]: Can you re-use orjson, or does it really have to be ujson?

File: scripts/chat.py [hunk 4]
```diff
@@ -63,10 +64,9 @@ def chat_with_ai(
             """
             model = cfg.fast_llm_model # TODO: Change model from hardcode to argument
             # Reserve 1000 tokens for the response
-            
-            if cfg.debug:
+            if os.environ.get("DEBUG"):
```
Comment (dhensen) [inline scripts/chat.py]: I think it's better to keep using `cfg.debug` and then move the env var to the `Config.debug` code. Or call `cfg.set_debug_mode(os.environ.get("DEBUG"))` even though those setter/getters are not pythonic.

**Review by dhensen — APPROVED** [review]

**Comment by Wladastic** [comment]
Looks intriguing, does it work with gpt3.5 at all?
I feel like it may require too many tokens at once.
Any plans to use Anaconda or Vicuna with it?
I just received a new GPU to run it locally.

**Comment by dhensen** [comment]
> Looks intriguing, does it work with gpt3.5 at all? I feel like it may require too many tokens at once. Any plans to use Anaconda or Vicuna with it? I just received a new GPU to run it locally.

Vicuna is really impressive... I don't have it running locally yet, no time yet. Would love to see/make a Vicuna integration for AutoGPT.

**Comment by merwanehamadi** [comment]
@Wladastic don't think tokens cost because I am telling you: in a few months you're going to have a bunch of things free, or at least very cheap, for the same quality of GPT3.5 and that will run locally (Vicuna is already great for chat and questions/answers, not so much to code, I have found)

I only develop with GPT 3.5 right now, because GPT4 will be too expansive to run CI/CD on.
The ability to change temperature and the "NoMemory" option will greatly increase the consistency of GPT3.5.
GPT 3.5 is completely overwhelmed by the memory provided in a json. It dilutes its answer, it forgets the initial goals it has. And the temperature set to 0 reduced drastically the json parsed errors. 

@dhensen yes, the only problem with Vicuna is it's 20Gb. If we have a github pipeline with it, we will have to create an instance in a cloud provider and call it from there, so it might be more expensive in the long run, because on github action it's going to be too much I think. But we might try.

Honestly, GPT3.5 is great and extremely cheap. I think it's a good middle-ground for now.

**Comment by merwanehamadi** [comment]
damn it, there is still an issue in my PR because the project is working at the scripts level.
@Torantulino we're going to need something like what @ryanpeach did. I am not going to do it here in this PR, because it's going to be too much changes at once. I will try and push something targeted.
the change is very simple: you keep everything except that scripts is now a module and you run scripts module.
Then after that we will rename scripts to auto-gpt. I would rename it directly but I know people have problems with big changes in this project... (however big changes are necessary here, scripts is not a good module name)

**Comment by nponeccop** [comment]
Make it a compound PR where the parts are more or less independent and can be easily reviewed. I don't want you to do anything, it's @Torantulino who decides. I'm sorry if I sounded like the official opinion of the repo. I'm not.

**Comment by Wladastic** [comment]
@merwanehamadi  Have you tried vicuna?
I am not even sure which model is best for coding help.
I want to mix gpt3 and vicuna, because gpt3 has goldfish memory and keeps repeating the same 4-5 steps for me...

**Review by jaspythonis — APPROVED** [review]

**Comment by nponeccop** [comment]
@merwanehamadi There are conflicts. But frankly you should only resolve them when you get a @Torantulino confirmation of merge.

**Comment by merwanehamadi** [comment]
Changes:
- Renamed autonomous_agents to agents, it's more generic, some agents are autonomous, some semi, some manual.
- swagger UI to document automatically the API
Screenshot 2023-04-11 at 6 23 01 PM


@Torantulino @nponeccop how about I just merge the "create agent endpoint" in a separate PR ? This endpoint will allow:
- to create a new agent. It will simply overwrite the existing one for now because we don't have many ai_settings.
- after this is merged I can push the endpoint to start and stop the agent from a web browser. For now this option will just create autonomous agents, because interacting with the agents requires other endpoints.

**Comment by merwanehamadi** [comment]
FYI @Torantulino a lot of the 56 files in this PR are tests

**Comment by nponeccop** [comment]
It's not about tests, it's about implementing multiple features at once. A single feature is fine, we don't count lines. But your PR is clearly splittable into small improvements that can be applied before you add the REST API.

**Comment by merwanehamadi** [comment]
OK I see, so here is my suggestion. I am making these code changes right now:
- allow to set temperature to 0
- have a memory that always retrieves nothing.
- wrap the main.py in a "if __main__ "
- wrap the while true in a method that can be reused

+ add the endpoints for the lifecycle of an agent: create, start and stop agent, along with integration tests.

How about I create a PR that just does this:
- add the "create agent" endpoint. This PR will setup a proper web server repo structure (with versioning of the API, which is super important) and allow the creation of an agent. And that's it. This PR will have very little change to the original code.

**Comment by merwanehamadi** [comment]
rebase is still ongoing. I prefer to rebase often, it's easier.
Just a little word on scripts.main.py:
- the while true loop really needs to go in a method.
- we need to stop using global, there is a new global logging that has been added 😄. When you need global it's the sign you need a class.

@Torantulino if I create a PR to do the items above, will you consider it ?

**Comment by merwanehamadi** [comment]
@nponeccop @Torantulino ok let's split the changes so that this is more digestible: 
- this one is huge: https://github.com/Torantulino/Auto-GPT/pull/1028 my agent is way better with low temperature

**Comment by merwanehamadi** [comment]
I have added the smallest change possible in this repo:
- web server with create agent endpoint: https://github.com/Torantulino/Auto-GPT/pull/1139

**Comment by Pwuts** [comment]
Resolves #70

**Comment by p-i-** [comment]
This is a mass message from the AutoGPT core team.
Our apologies for the ongoing delay in processing PRs.
This is because we are re-architecting the AutoGPT core!

For more details (and for infor on joining our Discord), please refer to:
  https://github.com/Significant-Gravitas/Auto-GPT/wiki/Architecting

**Comment by Boostrix** [comment]
the interesting thing here is that this sort of interface would make scaling rather straightforward - you could even hook up multiple Auto-GPT instances together, running on different computers in the same network/LAN or possibly by coming up with a clustered network where people could volunteer their Auto-GPT instances to work towards a common goal (think SETI/protein folding etc) - so this sort of feature could become the foundation for crowd computing, and the scale would be massive if each Auto-GPT has access to the same network and different LLMs - the interesting thing being that a distributed setup would add network latencies to the pool of resources to be tracked (#3466 )

**Comment by Boostrix** [comment]
The REST API would actually be good to have, as it would make it much easier to add a simple UI/web UI on top of the whole thing, without having to dissect all the underlying code.
poc-agpt-ui

**Comment by merwanehamadi** [comment]
a rest api is coming in the rearch

**Pull request closed without merging**